main/src/ca/uwaterloo/flix/language/phase/Lowering.scala

```diff
@@ -1242,6 +1309,71 @@ object Lowering {
     mkTag(Enums.HeadTerm, s"App$arity", innerExp, Types.HeadTerm, loc)
   }
 
+  /**
+    * Make the array of MpmcAdmin objects which will be passed to `selectFrom`
+    */
+  private def mkChannelAdminArray(rs: List[SelectChannelRule], channels: List[(Symbol.VarSym, Expression)], loc: SourceLocation): Expression = {
+    val admins = rs.zip(channels) map {
+      case (SelectChannelRule(_, c, _), channel) =>
+        val admin = Expression.Def(Defs.ChannelMpmcAdmin, Type.mkPureArrow(c.tpe, Types.ChannelMpmcAdmin, loc), loc)
+        Expression.Apply(admin, List(Expression.Var(channel._1, c.tpe, loc)), Types.ChannelMpmcAdmin, Type.Pure, Type.Empty, loc)
+    }
+    mkArray(admins, Types.ChannelMpmcAdmin, loc)
+  }
+
+  /**
+    * Construct a call to `selectFrom` given an array of MpmcAdmin objects and optional default
+    */
+  private def mkChannelSelect(adminArray: Expression, default: Option[Expression], loc: SourceLocation): Expression = {
+    val locksType = Types.mkList(Types.ConcurrentReentrantLock, loc)
+
+    val selectRetTpe = Type.mkTuple(List(Type.Int32, locksType), loc)
+    val selectTpe = Type.mkImpureUncurriedArrow(List(adminArray.tpe, Type.Bool), selectRetTpe, loc)
+    val select = Expression.Def(Defs.ChannelSelectFrom, selectTpe, loc)
+    val blocking = default match {
+      case Some(_) => Expression.False(loc)
+      case None => Expression.True(loc)
+    }
+    Expression.Apply(select, List(adminArray, blocking), selectRetTpe, Type.Impure, Type.Empty, loc)
+  }
+
+  /**
+    * Construct a sequence of MatchRules corresponding to the given SelectChannelRules
+    */
+  private def mkChannelCases(rs: List[SelectChannelRule], channels: List[(Symbol.VarSym, Expression)], retTpe: Type, pur: Type, eff: Type, loc: SourceLocation)(implicit flix: Flix): List[MatchRule] = {
+    val locksType = Types.mkList(Types.ConcurrentReentrantLock, loc)
+
+    rs.zip(channels).zipWithIndex map {
+      case ((SelectChannelRule(sym, chan, exp), channel), i) =>
+        val locksSym = mkLetSym("locks", loc)
+        val pat = Pattern.Tuple(List(Pattern.Int32(i, loc), Pattern.Var(locksSym, locksType, loc)), retTpe, loc)
```

Comment:
I was confused about whether using `retTpe` for the pattern was correct because I thought retTpe was the return type of the select. But actually its the type of the select expression (which is (Int32, LocksType)).

Reply:
I think it would be better if instead of taking retTpe as an argument we construct the argument here.

Thus I propose to introduce a helper function `mkPairExp(exp1, exp2, loc): Exp` which builds the `Pattern.Tuple` (and then uses the types of exp1 and exp2 to construct the right tuple type.